File: leetcode/solutions/python/problem_0001.py
from typing import List, Dict


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # O(n^2)
        # for first_idx in range(len(nums)):
        # for second_idx in range(first_idx+1, len(nums)):
        # if (nums[first_idx] + nums[second_idx]) == target:
        # return [first_idx, second_idx]

        # O(n)
        occurrences: Dict[int, int] = {}

        for idx, num in enumerate(nums):
            if target - num in occurrences:
                return [occurrences[target - num], idx]

            occurrences[num] = idx

File: leetcode/solutions/rust/src/problem_0001.rs
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        for (first_num_idx, first_num) in nums.iter().enumerate() {
            for (second_num_idx, second_num) in nums.iter().enumerate().skip(first_num_idx + 1) {
                if first_num + second_num == target {
                    return vec![first_num_idx as i32, second_num_idx as i32];
                }
            }
        }

        vec![]
    }
}
